src/jit/lowerxarch.cpp

```diff
@@ -35,6 +35,50 @@ void Lowering::LowerRotate(GenTreePtr tree)
 }
 
 //------------------------------------------------------------------------
+// LowerShift: Lower shift nodes
+//
+// Arguments:
+//    shift - the shift node (GT_LSH, GT_RSH or GT_RSZ)
+//
+// Notes:
+//    Remove unnecessary shift count masking, xarch shift instructions
+//    mask the shift count to 5 bits (or 6 bits for 64 bit operations).
+
+void Lowering::LowerShift(GenTreeOp* shift)
+{
+    assert(shift->OperIs(GT_LSH, GT_RSH, GT_RSZ));
+
+    size_t mask = 0x1f;
+#ifdef _TARGET_AMD64_
+    if (varTypeIsLong(shift->TypeGet()))
+    {
+        mask = 0x3f;
+    }
+#else
+    assert(!varTypeIsLong(shift->TypeGet()));
+#endif
+
+    for (GenTree* andOp = shift->gtGetOp2(); andOp->OperIs(GT_AND); andOp = andOp->gtGetOp1())
+    {
```

Comment:
> Why is this written as a for loop.

To catch multiple `and` instructions. See the PR description:
> It's also possible that developers also mask the shift count, not knowing that the C# language specification guarantees masking. In that case we end up with 2 and x, 31 instructions.

> This will match expressions that are not appropriate such as

Hmm, how so? The loop condition is `andOp->OperIs(GT_AND)` so it will stop at the first node that isn't an `and`.